### PR TITLE
11.0 mig base exception

### PR DESCRIPTION
**Pull request by mourad-ehm**
MIG module base_exception and add test.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Please squash some commits according https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests

**Comment by mourad-ehm** [comment]
Runbot is red but in runbot log I had a no error.  Someone know why runbot is red ? 
ping @pedrobaeza @jbeficent 
log message:

`
2017-10-16 15:27:52,164 151 INFO openerp_test odoo.addons.base.module.module: ALLOW access to module.button_install on ['base_exception'] to user admin #1 via n/a
2017-10-16 15:27:52,278 151 INFO openerp_test odoo.modules.loading: loading 12 modules...
2017-10-16 15:27:52,346 151 INFO openerp_test odoo.modules.loading: 12 modules loaded in 0.07s, 0 queries
2017-10-16 15:27:52,349 151 INFO openerp_test odoo.modules.loading: loading 13 modules...
2017-10-16 15:27:52,572 151 WARNING openerp_test odoo.fields: Field base.exception.test.model.line.lead_id with unknown comodel_name 'base.exception.test.model'
2017-10-16 15:27:52,761 151 INFO openerp_test odoo.modules.registry: module base_exception: creating or updating database tables
2017-10-16 15:27:52,863 151 INFO openerp_test odoo.models: Storing computed values of base.exception.test.purchase.main_exception_id
2017-10-16 15:27:52,872 151 INFO openerp_test odoo.models: Storing computed values of base.exception.test.purchase.amount_total
2017-10-16 15:27:53,028 151 INFO openerp_test odoo.modules.loading: loading base_exception/security/base_exception_security.xml
2017-10-16 15:27:53,139 151 INFO openerp_test odoo.modules.loading: loading base_exception/security/ir.model.access.csv
2017-10-16 15:27:53,233 151 INFO openerp_test odoo.modules.loading: loading base_exception/wizard/base_exception_confirm_view.xml
2017-10-16 15:27:53,262 151 INFO openerp_test odoo.modules.loading: loading base_exception/views/base_exception_view.xml
2017-10-16 15:27:53,352 151 INFO openerp_test odoo.modules.loading: 13 modules loaded in 1.00s, 0 queries
2017-10-16 15:27:53,513 151 INFO openerp_test odoo.modules.loading: Modules loaded.
2017-10-16 15:27:53,515 151 INFO openerp_test odoo.models: Computing parent left and right for table ir_ui_menu...
2017-10-16 15:27:53,532 151 INFO openerp_test odoo.service.server: Initiating shutdown
2017-10-16 15:27:53,533 151 INFO openerp_test odoo.service.server: Hit CTRL-C again or send a second signal to force the shutdown.
----------
Module test summary
[1;32mSuccess[0;m base_exception

`+`=======================================
|  Tests summary:
|---------------------------------------
| test_server.py              [1;32mSuccess[0;m
+=======================================
`

**Comment by yajo** [comment]
Runbot had some problems recently, I just rebuilt it, let's see.

**Comment by mourad-ehm** [comment]
I rebuild runbot and had this error:
Unable to find image 'oca-server-tools:8bfb488_2' locally
docker: Error response from daemon: pull access denied for oca-server-tools, repository does not exist or may require 'docker login'.
See 'docker run --help'.

**Comment by pedrobaeza** [comment]
It seems to be a problem with runbot. I have notified already @gurneyalex about it.

**Comment by mourad-ehm** [comment]
Now runbot is build correctly but is still red, someone know why ? ping  @gurneyalex @legalsylvain

**Comment by pedrobaeza** [comment]
See logs. You will find this warning:
```
2017-10-24 13:29:09,284 151 WARNING openerp_test odoo.fields: Field base.exception.test.model.line.lead_id with unknown comodel_name 'base.exception.test.model'
```

**Comment by JordiBForgeFlow** [comment]
I have proposed small fixes here: https://github.com/Eficent/server-tools/tree/11.0-mig-base_exception

But cannot make a PR, the PR does not allow me to select the Akretion repo.

**Comment by mourad-ehm** [comment]
Thanks @jbeficent. I integrated your commit.

**Comment by JordiBForgeFlow** [comment]
@mourad-ehm I still added another commit https://github.com/Eficent/server-tools/commit/b66bddf817f66bf121af9e0afba328816ebead2e in https://github.com/Eficent/server-tools/tree/11.0-mig-base_exception.

**Comment by bealdav** [comment]
Here codecov is 55% but other tests coverage in purchase_exception will cover other parts of the code of this module. Then I think it's mergeable as it. Thanks @mourad-ehm for this work

**Review by bealdav — APPROVED** [review]

File: base_exception/models/base_exception.py
Comment (hparfr) [inline base_exception/models/base_exception.py]: set the state

File: base_exception/models/base_exception.py
Comment (hparfr) [inline base_exception/models/base_exception.py]: `block = True` or `failed = True`  (code.default)

File: base_exception/models/base_exception.py
Comment (hparfr) [inline base_exception/models/base_exception.py]: put comprehension list out of `if` clause to add readability

File: base_exception/models/base_exception.py
Comment (hparfr) [inline base_exception/models/base_exception.py]: merge in one line or use different variable names

File: base_exception/models/base_exception.py
Comment (hparfr) [inline base_exception/models/base_exception.py]: old api ?

File: base_exception/models/base_exception.py
Comment (hparfr) [inline base_exception/models/base_exception.py]: useless next_state_rule condition (because of else)

Comment (hparfr) [reply base_exception/models/base_exception.py]:
```py
 if not next_state_rule or rule.sequence < next_state_rule.sequence
  next_state_rule = rule
```

File: base_exception/wizard/base_exception_confirm.py
Comment (hparfr) [inline base_exception/wizard/base_exception_confirm.py]: raise instead of assert

**Review by hparfr — COMMENTED** [review]
Some remarks and questions

File: base_exception/models/base_exception.py
Comment (mourad-ehm) [reply base_exception/models/base_exception.py]: No, this is not old API but variables which can use in Python code with safe_eval

Comment (rousseldenis) [reply base_exception/models/base_exception.py]: @mourad-ehm self.pool is not meant to be used.

I see it was migrated in v10 like that  but it's incorrect. Use self.env[rec._name] and 'env': self.env instead of pool.

But I think changing function to @api.multi and returning {'self': self} should be sufficent. So, you can change the function signature removing obj_name and rec parameters!

File: base_exception/models/base_exception.py
Comment (rousseldenis) [inline base_exception/models/base_exception.py]: user = self.env.user

File: base_exception/models/base_exception.py
Comment (rousseldenis) [inline base_exception/models/base_exception.py]: self.env.user.id

File: base_exception/models/base_exception.py
Comment (rousseldenis) [inline base_exception/models/base_exception.py]: self.env.cr

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Code review

File: base_exception/models/base_exception.py
Comment (bealdav) [reply base_exception/models/base_exception.py]: I thought `self._cr`, `self._uid` were shortcuts for self.env.cr, etc.
Are they now deprecated ?
Thanks for your review

**Comment by mourad-ehm** [comment]
Thanks @rousseldenis  for your review. I fixed your remarks.

File: base_exception/models/base_exception.py
Comment (rousseldenis) [reply base_exception/models/base_exception.py]: @bealdav See https://github.com/odoo/odoo/blob/ecbf7aa4714479229658d14cce28fa00376ed390/odoo/models.py#L4254

They are kept for backward compatibility but their future is not sure.

File: base_exception/models/base_exception.py
Comment (rousseldenis) [inline base_exception/models/base_exception.py]: @mourad-ehm I forget self.env.context

File: base_exception/models/base_exception.py
Comment (rousseldenis) [inline base_exception/models/base_exception.py:227]: @mourad-ehm IMHO Same as above, you could decorate your _rule_eval function with api.multi and remove both self.rule_group and self arguments.

File: base_exception/wizard/base_exception_confirm.py
Comment (rousseldenis) [inline base_exception/wizard/base_exception_confirm.py]: self.env.context

File: base_exception/wizard/base_exception_confirm.py
Comment (rousseldenis) [inline base_exception/wizard/base_exception_confirm.py]: Same as above

**Review by rousseldenis — CHANGES_REQUESTED** [review]
Some remarks

File: base_exception/models/base_exception.py
Comment (bealdav) [reply base_exception/models/base_exception.py]: thanks a lot

File: base_exception/wizard/base_exception_confirm_view.xml
Comment (simahawk) [inline base_exception/wizard/base_exception_confirm_view.xml]: fix indentation

File: base_exception/wizard/base_exception_confirm.py
Comment (simahawk) [inline base_exception/wizard/base_exception_confirm.py]: No need to loop, just: `related_model_except.exception_ids.ids`

File: base_exception/__init__.py
Comment (simahawk) [inline base_exception/__init__.py:5]: if you import it here w/out any condition is going to always be available.
Usualy I wrap the code of the test model inside an IF block. Like:
```python
testing = tools.config.get('test_enable') or os.environ.get('ODOO_TEST_ENABLE')
if testing:
    class Foo(...):
[...]
```
In this way is available only when running tests or on demand w/ the env var.

**Review by simahawk — CHANGES_REQUESTED** [review]

File: base_exception/__init__.py
Comment (mourad-ehm) [reply base_exception/__init__.py:5]: Thanks for review. What's about import of ir.model.access.csv in the installation with testing = false ?

Comment (simahawk) [reply base_exception/__init__.py:5]: don't care about that, remove ACL for this modeel. Is a testing model used for tests.

Comment (pedrobaeza) [reply base_exception/__init__.py:5]: Can you test it with an existing base model instead of creating one?

Comment (simahawk) [reply base_exception/__init__.py:5]: @pedrobaeza actually if you want to simulate real usage is better to define a test model.
@mourad-ehm I just fixed an issue related to this w/ my approach here https://github.com/OCA/partner-contact/pull/526/commits/04a5bb7280238da000fd2dd764215cf257e904c8

File: base_exception/models/base_exception.py
Comment (mourad-ehm) [reply base_exception/models/base_exception.py:227]: I can't do this because I use the same method for object and lines. 
code: if self._rule_eval(rule, group_line, obj_line):

**Review by hparfr — APPROVED** [review]

File: base_exception/__init__.py
Comment (mourad-ehm) [reply base_exception/__init__.py:5]: But travis test is fail because of ACL https://travis-ci.org/OCA/server-tools/jobs/309585651#L1032

**Comment by mourad-ehm** [comment]
Hi @moylop260, I get this error in travis test:
************* Module base_exception.models.base_exception

base_exception/models/base_exception.py:1: [C8201(no-utf8-coding-comment), ] No UTF-8 coding comment found: Use `# coding: utf-8` or `# -*- coding: utf-8 -*-`

In v 11 we use python 3 and it's not necessary. The tests don't fail 3 days ago.

**Comment by moylop260** [comment]
@mourad-ehm I got it thank you we will followup it from https://github.com/OCA/pylint-odoo/issues/164

**Comment by mourad-ehm** [comment]
Thanks @moylop260 for fix.

File: base_exception/__init__.py
Comment (mourad-ehm) [reply base_exception/__init__.py:5]: ping @simahawk. I added temp ACL to validate travis test. https://github.com/OCA/server-tools/pull/1025/commits/785be4dfa2ce767100f4fdbc3adb2e01b01c9956

File: base_exception/models/base_exception.py
Comment (mourad-ehm) [reply base_exception/models/base_exception.py:227]: ping @rousseldenis

**Comment by mourad-ehm** [comment]
Thanks @jbeficent, I integrated your commits.

File: base_exception/models/base_exception.py
Comment (rousseldenis) [reply base_exception/models/base_exception.py:227]: ok!

Comment (mourad-ehm) [reply base_exception/models/base_exception.py:227]: You can approve :)

Comment (rousseldenis) [reply base_exception/models/base_exception.py:227]: oups. For sure

**Review by rousseldenis — APPROVED** [review]
Code review.
Should add more tests on roadmap to improve coverage.

**Comment by mourad-ehm** [comment]
OK, thanks @rousseldenis

File: base_exception/__init__.py
Comment (simahawk) [reply base_exception/__init__.py:5]: what you are linking is the std odoo warning and if you look into that build the LINT tests is green. In any case, I think you can disable MQT check for ACL on that file via `# pylint disable: $something`.  @moylop260?

Comment (moylop260) [reply base_exception/__init__.py:5]: The tests folder is imported by default from odoo by standard `module/tests` is not required explicit import from `__init__.py`

Comment (moylop260) [reply base_exception/__init__.py:5]: I'm seeing the last build of this PR but I saw a green result from travis.
What is the message line (link) to silent it?

Comment (mourad-ehm) [reply base_exception/__init__.py:5]: Travis is green because I added this fix https://github.com/OCA/server-tools/pull/1025/commits/785be4dfa2ce767100f4fdbc3adb2e01b01c9956 .
In fact travis require access.model for temporary models used for test.

Comment (mourad-ehm) [reply base_exception/__init__.py:5]: Hi @simahawk @moylop260 , I think we can merge this PR (travis and runbot are green) and make other PR if necessary to drop ACL data.

Comment (simahawk) [reply base_exception/__init__.py:5]: fine for me. Just drop a line in ROADMAP so that we don't forget it. Thanks

**Review by simahawk — APPROVED** [review]

**Comment by simahawk** [comment]
@mourad-ehm remember to squash ;)

**Review by moylop260 — APPROVED** [review]

**Comment by mourad-ehm** [comment]
Ok, thanks @simahawk. I done it.

**Comment by mourad-ehm** [comment]
Hi @pedrobaeza, could you merge this PR ?

**Comment by pedrobaeza** [comment]
Can you please squash commits by author/logic set before merging?

**Comment by mourad-ehm** [comment]
Squash done.

**Comment by bealdav** [comment]
@pedrobaeza thanks for the merge. No project in v11 at akretion for now.

It's just to help the community. Thanks

**Comment by pedrobaeza** [comment]
Thank you! Much appreciated.

Merging.

File: base_exception/__init__.py
Comment (simahawk) [reply base_exception/__init__.py:5]: >The tests folder is imported by default from odoo by standard module/tests is not required explicit import from __init__.py

@moylop260 is this always true? I fear that is not imported unless you are specifically running tests for this module. [sale_exception PR](https://github.com/OCA/sale-workflow/pull/576) is failing because ["Model base.exception.test.purchase.line has no table"](https://travis-ci.org/OCA/sale-workflow/jobs/311944678#L1317)

Sounds like the models are registered at setup but they are not there anymore when running tests for sale_exception.

**Pull request merged**
